### PR TITLE
Improved evaluation creation

### PR DESCRIPTION
**Pull request by mehalshah**
Dynamic table to enable creation of PLC Evaluation Question and Plc Evaluation Answers.

Diagram of relationship of PLC Objects - http://wiki.code.org/display/Operations/Explanation+of+PLC+Model

### CODE REVIEW

File: code-studio/src/js/plc/evaluation_creation.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
+/* global dashboard */
+
+$(window).load(function () {
+  $('#evaluationTable .new_question').click(function (event) {
+    var bottomRow = event.target.parentElement.parentElement;
+    var clonedRow = $('#evaluationTable .cloneableQuestionRow')[0].cloneNode(true);
+    clonedRow.className = 'new_question_row';
+    $(clonedRow).find('input')[0].className = 'new_question_name';
```
Comment (mehalshah) [inline code-studio/src/js/plc/evaluation_creation.js]: Use Jquery

File: code-studio/src/js/plc/evaluation_creation.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
+/* global dashboard */
+
+$(window).load(function () {
+  $('#evaluationTable .new_question').click(function (event) {
+    var bottomRow = event.target.parentElement.parentElement;
+    var clonedRow = $('#evaluationTable .cloneableQuestionRow')[0].cloneNode(true);
```
Comment (mehalshah) [inline code-studio/src/js/plc/evaluation_creation.js]: Use Jquery

File: code-studio/src/js/plc/evaluation_creation.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    clonedRow.className = 'new_question_row';
+    $(clonedRow).find('input')[0].className = 'new_question_name';
+
+    $(clonedRow).insertBefore(bottomRow);
+
+    $('#evaluationTable input').last().keyup(handleNewQuestionName);
+  });
+
+  $('#evaluationTable .new_answer').click(function (event) {
+    var clonedRow = $(event.target.parentElement.parentElement.parentElement).find('.cloneableAnswerRow')[0].cloneNode(true);
```
Comment (mehalshah) [inline code-studio/src/js/plc/evaluation_creation.js]: Use Jquery

File: code-studio/src/js/plc/evaluation_creation.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    clonedRow.removeClass('cloneableAnswerRow');
+    clonedRow.addClass('new_answer_row');
+    clonedRow.attr('question_id', event.target.getAttribute('question_id'));
+    clonedRow.find('input').keyup(handleNewAnswer);
+    clonedRow.find('select').change(handleNewAnswer);
+    clonedRow.insertBefore(event.target.parentElement.parentElement);
+  });
+
+  function handleNewQuestionName() {
+    $('#newQuestionsList').val(JSON.stringify($('.new_question_row .new_question_name').map(function() { return $(this).val(); }).get()));
```
Comment (Bjvanminnen) [inline code-studio/src/js/plc/evaluation_creation.js]: style nit: would prefer this to be on multiple lines

```
$('#newQuestionsList').val(JSON.stringify($('.new_question_row .new_question_name').map(function() {
  return $(this).val();
}).get()))
```

File: code-studio/src/js/plc/evaluation_creation.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+    $('.question_section .new_answer_row').each(function (index, element) {
+      var questionId = element.getAttribute('question_id');
+      var answerText = $(element).find('input').val();
+      var associatedModule = $(element).find('select').val();
+
+      if(newAnswerData[questionId]) {
+        newAnswerData[questionId].push({'answer': answerText, 'learningModuleId': associatedModule});
+      } else {
+        newAnswerData[questionId] = [{'answer': answerText, 'learningModuleId': associatedModule}];
```
Comment (Bjvanminnen) [inline code-studio/src/js/plc/evaluation_creation.js]: Style nit: No need for keys to be quoted, unless they have special characters in them. Also, style preference is for objects with multiple keys to get split per line, i.e.

```
newAnswerData[questionId] = [{
  answer: answerText,
  learningModuleId: associatedModule
}];
```

File: code-studio/src/js/plc/evaluation_creation.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+      }
+    });
+
+    $('#newAnswersList').val(JSON.stringify(newAnswerData));
+    $('#submitNewQuestions').prop('disabled', false);
+  }
+});
```
Comment (Bjvanminnen) [inline code-studio/src/js/plc/evaluation_creation.js:53]: Out of curiosity, was there any consideration for doing this in React?

File: code-studio/src/js/plc/evaluation_creation.js (hunk 5 above)
Comment (Bjvanminnen) [inline code-studio/src/js/plc/evaluation_creation.js]: You could also do things this way, avoiding the conditional and having only one line where you construct the format for your objects:

```
newAnswerData[questionId] = newAnswerData[questionId] || [];
newAnswerData[questionId].push({
  answer: answerText,
  learningModuleId: associatedModule
});
```

File: dashboard/app/views/plc/courses/show.html.haml [hunk 7]
```diff
@@ -1,6 +1,67 @@
...
+  var allModuleNames = #{options_for_plc_task_learning_modules}
+
 %p
   %strong Name: #{@course.name}
 
+%p
+  Evaluation Questions
+
+%p
+  %table#evaluationTable{ style: 'width: 100%' }
```
Comment (Bjvanminnen) [inline dashboard/app/views/plc/courses/show.html.haml]: This style is redundant, as it's also declared in our css right?

File: dashboard/app/views/plc/courses/show.html.haml [hunk 8]
```diff
@@ -1,6 +1,67 @@
+- content_for(:head) do
+  = stylesheet_link_tag 'css/levelbuilder', media: 'all'
+  = javascript_include_tag 'js/plc'
+
+
+:javascript
+  var allModuleNames = #{options_for_plc_task_learning_modules}
```
Comment (Bjvanminnen) [inline dashboard/app/views/plc/courses/show.html.haml]: I dont see `allModuleNames` used anywhere?

File: code-studio/src/js/plc/evaluation_creation.js (hunk 4 above)
Comment (mehalshah) [inline code-studio/src/js/plc/evaluation_creation.js]: Done

File: code-studio/src/js/plc/evaluation_creation.js (hunk 5 above)
Comment (mehalshah) [inline code-studio/src/js/plc/evaluation_creation.js]: Done

File: code-studio/src/js/plc/evaluation_creation.js (hunk 6 above)
Comment (mehalshah) [inline code-studio/src/js/plc/evaluation_creation.js:53]: Didn't really cross my mind - I'm not that familiar with it and didn't want to ramp up on something new for a short project

File: dashboard/app/views/plc/courses/show.html.haml (hunk 7 above)
Comment (mehalshah) [inline dashboard/app/views/plc/courses/show.html.haml]: Yup - will remove

File: dashboard/app/views/plc/courses/show.html.haml (hunk 8 above)
Comment (mehalshah) [inline dashboard/app/views/plc/courses/show.html.haml]: Ah - this was part of a scheme to get this into the evaluation_creation.js but it isn't needed anymore - will remove it

File: code-studio/src/js/plc/evaluation_creation.js (hunk 4 above)
Comment (mehalshah) [inline code-studio/src/js/plc/evaluation_creation.js]: Done

File: code-studio/src/js/plc/evaluation_creation.js (hunk 5 above)
Comment (mehalshah) [inline code-studio/src/js/plc/evaluation_creation.js]: Done

**Pull request merged**